src/main/java/graphql/schema/transform/FieldVisibilitySchemaTransformation.java

```diff
@@ -0,0 +1,137 @@
+
+        return finalSchema;
+    }
+
+    private static class FieldVisibilityVisitor extends GraphQLTypeVisitorStub {
+
+        private final VisibleFieldPredicate visibilityPredicate;
+        private final Set<GraphQLType> removedTypes;
+        private final Set<GraphQLType> observedTypes;
+
+        private FieldVisibilityVisitor(VisibleFieldPredicate visibilityPredicate,
+                                       Set<GraphQLType> removedTypes, Set<GraphQLType> observedTypes) {
+            this.visibilityPredicate = visibilityPredicate;
+            this.removedTypes = removedTypes;
+            this.observedTypes = observedTypes;
+        }
+
+        @Override
+        public TraversalControl visitGraphQLInterfaceType(GraphQLInterfaceType node,
+                                                          TraverserContext<GraphQLSchemaElement> context) {
+            observedTypes.add(node);
+
+            return TraversalControl.CONTINUE;
+        }
+
+        @Override
+        public TraversalControl visitGraphQLObjectType(GraphQLObjectType node,
+                                                       TraverserContext<GraphQLSchemaElement> context) {
+
+            if (context.getBreadcrumbs().stream()
+                    .anyMatch(crumb -> crumb.getLocation().getName().equalsIgnoreCase("addTypes"))) {
+                return TraversalControl.ABORT;
+            }
+
+            observedTypes.add(node);
+
+            return TraversalControl.CONTINUE;
+        }
+
+        @Override
+        public TraversalControl visitGraphQLFieldDefinition(GraphQLFieldDefinition definition,
+                                                            TraverserContext<GraphQLSchemaElement> context) {
+            VisibleFieldPredicateEnvironment environment = new VisibleFieldPredicateEnvironmentImpl(
+                    definition, context.getParentNode());
+            if (!visibilityPredicate.isVisible(environment)) {
+                deleteNode(context);
+                removedTypes.add(definition.getType());
+            }
+
+            return TraversalControl.CONTINUE;
+        }
+    }
+
+    private static class TypeVisibilityVisitor extends GraphQLTypeVisitorStub {
+
+        private final String queryTypeName;
+        private final Set<GraphQLType> observedTypes;
+        private final Set<GraphQLType> removedTypes;
+
+        private TypeVisibilityVisitor(String queryTypeName,
+                                      Set<GraphQLType> observedTypes,
+                                      Set<GraphQLType> removedTypes) {
+            this.queryTypeName = queryTypeName;
+            this.observedTypes = observedTypes;
+            this.removedTypes = removedTypes;
+        }
+
+        @Override
+        public TraversalControl visitGraphQLObjectType(GraphQLObjectType node,
+                                                       TraverserContext<GraphQLSchemaElement> context) {
+            if (!observedTypes.contains(node) &&
```

Comment:
what about the Mutation and the Subscription root type? If you treat queryTypeName special you probably have to treat the others also accordingly

Reply:
We should also have tests for transforming schemas with mutation and subscriptions.

Reply:
Added.